onchain/permissionless-arbitration/offchain/player/strategy.lua

```diff
@@ -0,0 +1,253 @@
+            assert(f)
+        end
+
+        helper.log(player.player_index, string.format(
+            "advance match with current height %d in tournament %s of level %d for commitment %s",
+            match.current_height,
+            match.tournament.address,
+            match.tournament.level,
+            commitment.root_hash
+        ))
+        local ok, e = pcall(player.client.tx_advance_match,
+            player.client,
+            match.tournament.address,
+            match.commitment_one,
+            match.commitment_two,
+            left,
+            right,
+            new_left,
+            new_right
+        )
+        if not ok then
+            helper.log(player.player_index, string.format(
+                "advance match reverted: %s",
+                e
+            ))
+        end
+    end
+end
+
+_react_tournament_honestly = function(player, tournament)
+    local tournament_winner = tournament.tournament_winner
+    if tournament_winner[1] == "true" then
+        if not tournament.parent then
+            helper.log(player.player_index, "TOURNAMENT FINISHED, HURRAYYY")
+            helper.log(player.player_index, "Winner commitment: " .. tournament_winner[2]:hex_string())
+            helper.log(player.player_index, "Final state: " .. tournament_winner[3]:hex_string())
+            return true
+        else
+            local old_commitment = tournament.parent.commitment
+            if tournament_winner[2] ~= old_commitment.root_hash then
+                helper.log(player.player_index, "player lost tournament")
+                player.has_lost = true
+                return
+            end
+
+            if tournament.called_win then
+                helper.log(player.player_index, "player already called winInnerMatch")
+                return
+            else
+                tournament.called_win = true
+            end
+
+            helper.log(player.player_index, string.format(
+                "win tournament %s of level %d for commitment %s",
+                tournament.address,
+                tournament.level,
+                tournament.commitment.root_hash
+            ))
+            local _, left, right = old_commitment:children(old_commitment.root_hash)
+            local ok, e = pcall(player.client.tx_win_inner_match,
+                player.client,
+                tournament.parent.address,
+                tournament.address,
+                left,
+                right
+            )
+            if not ok then
+                helper.log(player.player_index, string.format(
+                    "win inner match reverted: %s",
+                    e
+                ))
+            end
+            return
+        end
+    end
+
+    if not tournament.latest_match then
+        _join_tournament_if_needed(player, tournament)
+    else
+        _react_match_honestly(player, tournament.latest_match, tournament.commitment)
+    end
+end
+
+local function _react_honestly(player)
```

Comment:
Maybe we can turn this whole thing into this `HonestStrategy`. So all strategies implement the method `react`, but the honest one does it honestly. Maybe we can create a nice type for it.